File: 程序员面试金典/4.6.后继者.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        pre = None
        while p != root:
            if p.val > root.val:
                root = root.right
            else:
                pre = root
                root = root.left

        if root.right:
            root = root.right
            while root.left:
                root = root.left

            return root

        else:
            return pre

File: 程序员面试金典/4.6.后继者.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p)
    {
        TreeNode* pre = NULL;
        while (p != root) {
            if (p->val <= root->val) {
                pre = root;
                root = root->left;
            } else {
                root = root->right;
            }
        }

        if (p->right) {
            p = p->right;
            while (p->left) {
                p = p->left;
            }
            return p;
        } else {
            return pre;
        }
    }
};